File: crc_patcher.js
const fs = require('fs').promises;
const path = require('path'); 



function bytes_to_u32_be(b) {
    return b.readUInt32BE(0);
}

function u32_to_bytes_be(i) {
    const buf = Buffer.alloc(4);
    if (typeof i === 'bigint') {
        buf.writeUInt32BE(Number(i & 0xFFFFFFFFn), 0);
    } else {
        buf.writeUInt32BE(i, 0);
    }
    return buf;
}

function reverse_bits_in_bytes(b) {
    const num = bytes_to_u32_be(b);
    let rev = 0;
    for (let i = 0; i < 32; i++) {
        if ((num >> i) & 1) {
            rev |= 1 << (31 - i);
        }
    }
    return u32_to_bytes_be(rev >>> 0);
}

function gf_multiply(a, b) {
    let result = 0n;
    while (b > 0n) {
        if (b & 1n) {
            result ^= a;
        }
        a <<= 1n;
        b >>= 1n;
    }
    return result;
}

function gf_divide(dividend, divisor) {
    if (divisor === 0n) return 0n;
    let quotient = 0n;
    let remainder = dividend;
    const divisor_bits = BigInt(divisor.toString(2).length);
    while (remainder.toString(2).length >= divisor_bits && remainder !== 0n) {
        const shift = BigInt(remainder.toString(2).length) - divisor_bits;
        quotient |= 1n << shift;
        remainder ^= divisor << shift;
    }
    return quotient;
}

function gf_mod(dividend, divisor) {
    if (divisor === 0n) return dividend;
    const divisor_bits = BigInt(divisor.toString(2).length);
    while (dividend !== 0n && dividend.toString(2).length >= divisor_bits) {
        const shift = BigInt(dividend.toString(2).length) - divisor_bits;
        dividend ^= divisor << shift;
    }
    return dividend;
}

function gf_multiply_modular(a, b, modulus) {
    const product = gf_multiply(a, b);
    return gf_mod(product, modulus);
}

function gf_modular_inverse(a, m) {
    if (a === 0n) throw new Error("Inverse of zero does not exist");
    let [old_r, r] = [m, a];
    let [old_s, s] = [0n, 1n];
    while (r !== 0n) {
        const q = gf_divide(old_r, r);
        [old_r, r] = [r, old_r ^ gf_multiply(q, r)];
        [old_s, s] = [s, old_s ^ gf_multiply(q, s)];
    }
    if (old_r !== 1n) throw new Error("Modular inverse does not exist");
    return old_s;
}

function gf_inverse(k, poly) {
    const x32 = 0x100000000n;
    const inverse = gf_modular_inverse(x32, poly);
    return gf_multiply_modular(k, inverse, poly);
}

let crc32Table;

function makeCrc32Table() {
    crc32Table = new Uint32Array(256);
    for (let i = 0; i < 256; i++) {
        let c = i;
        for (let j = 0; j < 8; j++) {
            c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1);
        }
        crc32Table[i] = c;
    }
}

function compute_crc32(data) {
    if (!crc32Table) makeCrc32Table();
    let crc = 0xFFFFFFFF;
    for (let i = 0; i < data.length; i++) {
        crc = crc32Table[(crc ^ data[i]) & 0xFF] ^ (crc >>> 8);
    }
    return (crc ^ 0xFFFFFFFF) >>> 0;
}

function xor_bytes(a, b) {
    const length = Math.min(a.length, b.length);
    const result = Buffer.alloc(length);
    for (let i = 0; i < length; i++) {
        result[i] = a[i] ^ b[i];
    }
    return result;
}

async function manipulate_crc(original_path, mod_path, target_path = null) {
    try {
        // 如果只傳入兩個參數，保持舊的行為（向後兼容）
        if (target_path === null) {
            target_path = mod_path;
            mod_path = mod_path;
        }

        const original_data = await fs.readFile(original_path);
        const mod_data = await fs.readFile(mod_path);

        const original_crc = compute_crc32(original_data);
        const mod_crc = compute_crc32(mod_data);

        // 詳細顯示 CRC 信息用於調試
        console.log(`=== CRC DEBUG INFO ===`);
        console.log(`File: ${path.basename(target_path)}`);
        console.log(`Original file: ${original_path}`);
        console.log(`Mod file: ${mod_path}`);
        console.log(`Target file: ${target_path}`);
        console.log(`Original CRC: 0x${original_crc.toString(16).toUpperCase()} (${original_crc})`);
        console.log(`Mod CRC: 0x${mod_crc.toString(16).toUpperCase()} (${mod_crc})`);
        console.log(`Original file size: ${original_data.length} bytes`);
        console.log(`Mod file size: ${mod_data.length} bytes`);
        console.log(`=== END DEBUG INFO ===`);

        // 如果 mod 檔案和原始檔案的 CRC 相同，則直接複製
        if (mod_crc === original_crc) {
            console.log(`Mod CRC matches original CRC for "${path.basename(target_path)}". Copying without patch.`);
            await fs.writeFile(target_path, mod_data);
            return true;
        }

        console.log(`CRC mismatch for "${path.basename(target_path)}". Original: ${original_crc.toString(16)}, Mod: ${mod_crc.toString(16)}. Starting patch...`);

        const modified_crc_with_padding = compute_crc32(Buffer.concat([mod_data, Buffer.alloc(4)]));

        const original_bytes = u32_to_bytes_be(original_crc);
        const modified_bytes = u32_to_bytes_be(modified_crc_with_padding);

        const xor_result = xor_bytes(original_bytes, modified_bytes);
        const reversed_bytes = reverse_bits_in_bytes(xor_result);
        const k = BigInt(bytes_to_u32_be(reversed_bytes));

        const crc32_poly = 0x104C11DB7n;
        const correction_value = gf_inverse(k, crc32_poly);

        const correction_bytes_raw = u32_to_bytes_be(correction_value);
        const reverse_byte_bits = (byte) => parseInt(byte.toString(2).padStart(8, '0').split('').reverse().join(''), 2);
        const correction_bytes = Buffer.from([...correction_bytes_raw].map(b => reverse_byte_bits(b)));

        const final_data = Buffer.concat([mod_data, correction_bytes]);
        const final_crc = compute_crc32(final_data);

        if (final_crc === original_crc) {
            // 將修正後的數據寫入目標路徑
            await fs.writeFile(target_path, final_data);
            console.log(`Successfully patched CRC for "${path.basename(target_path)}". New CRC: ${final_crc.toString(16)}`);
            return true;
        }

        console.warn(`CRC patching FAILED for "${path.basename(target_path)}". Final CRC (${final_crc.toString(16)}) did not match original CRC (${original_crc.toString(16)}).`);
        return false;
    } catch (err) {
        console.error(`CRC manipulation process failed for ${target_path}:`, err);
        return false;
    }
}


module.exports = {
    manipulate_crc,
};

File: Blue Archive Mod Manager-C#/Utils/CrcPatcher.cs
using System;
using System.IO;
using System.Numerics;
using System.Threading.Tasks;

namespace Blue_Archive_Mod_Manager_C_.Utils
{
    public static class CrcPatcher
    {
        private static readonly uint[] Crc32Table = CreateCrc32Table();

        private static uint[] CreateCrc32Table()
        {
            var table = new uint[256];
            for (uint i = 0; i < 256; i++)
            {
                uint c = i;
                for (int j = 0; j < 8; j++)
                {
                    c = (c & 1) != 0 ? (0xEDB88320u ^ (c >> 1)) : (c >> 1);
                }
                table[i] = c;
            }
            return table;
        }

        private static uint ComputeCrc32(ReadOnlySpan<byte> data)
        {
            uint crc = 0xFFFFFFFF;
            foreach (var b in data)
            {
                crc = Crc32Table[(crc ^ b) & 0xFF] ^ (crc >> 8);
            }
            return crc ^ 0xFFFFFFFFu;
        }

        private static byte[] UInt32ToBytesBE(BigInteger value)
        {
            uint v = (uint)(value & 0xFFFFFFFF);
            return new[]
            {
                (byte)((v >> 24) & 0xFF),
                (byte)((v >> 16) & 0xFF),
                (byte)((v >> 8) & 0xFF),
                (byte)(v & 0xFF)
            };
        }

        private static uint BytesToUInt32BE(ReadOnlySpan<byte> b)
        {
            return (uint)(b[0] << 24 | b[1] << 16 | b[2] << 8 | b[3]);
        }

        private static byte[] ReverseBitsInBytes(ReadOnlySpan<byte> input)
        {
            uint num = BytesToUInt32BE(input);
            uint rev = 0;
            for (int i = 0; i < 32; i++)
            {
                if (((num >> i) & 1) == 1)
                {
                    rev |= 1u << (31 - i);
                }
            }
            return UInt32ToBytesBE(rev);
        }

        private static byte[] XorBytes(ReadOnlySpan<byte> a, ReadOnlySpan<byte> b)
        {
            int len = Math.Min(a.Length, b.Length);
            var result = new byte[len];
            for (int i = 0; i < len; i++)
            {
                result[i] = (byte)(a[i] ^ b[i]);
            }
            return result;
        }

        private static int BitLength(BigInteger value)
        {
            int bits = 0;
            var v = value;
            while (v > 0)
            {
                v >>= 1;
                bits++;
            }
            return bits;
        }

        private static BigInteger GfMultiply(BigInteger a, BigInteger b)
        {
            BigInteger result = 0;
            var x = a;
            var y = b;
            while (y > 0)
            {
                if ((y & 1) != 0) result ^= x;
                x <<= 1;
                y >>= 1;
            }
            return result;
        }

        private static BigInteger GfMod(BigInteger dividend, BigInteger divisor)
        {
            if (divisor == 0) return dividend;
            int divisorBits = BitLength(divisor);
            var d = dividend;
            while (d != 0 && BitLength(d) >= divisorBits)
            {
                int shift = BitLength(d) - divisorBits;
                d ^= divisor << shift;
            }
            return d;
        }

        private static BigInteger GfMultiplyModular(BigInteger a, BigInteger b, BigInteger modulus)
        {
            return GfMod(GfMultiply(a, b), modulus);
        }

        private static BigInteger GfDivide(BigInteger dividend, BigInteger divisor)
        {
            if (divisor == 0) return 0;
            BigInteger quotient = 0;
            BigInteger remainder = dividend;
            int divisorBits = BitLength(divisor);
            while (remainder != 0 && BitLength(remainder) >= divisorBits)
            {
                int shift = BitLength(remainder) - divisorBits;
                quotient |= BigInteger.One << shift;
                remainder ^= divisor << shift;
            }
            return quotient;
        }

        private static BigInteger GfModularInverse(BigInteger a, BigInteger modulus)
        {
            if (a == 0) throw new InvalidOperationException("Inverse of zero does not exist");

            BigInteger oldR = modulus, r = a;
            BigInteger oldS = 0, s = 1;

            while (r != 0)
            {
                var q = GfDivide(oldR, r);
                (oldR, r) = (r, oldR ^ GfMultiply(q, r));
                (oldS, s) = (s, oldS ^ GfMultiply(q, s));
            }

            if (oldR != 1) throw new InvalidOperationException("Modular inverse does not exist");
            return oldS;
        }

        private static BigInteger GfInverse(BigInteger k, BigInteger poly)
        {
            BigInteger x32 = 0x100000000;
            var inverse = GfModularInverse(x32, poly);
            return GfMultiplyModular(k, inverse, poly);
        }

        public static async Task<bool> ManipulateCrcAsync(string originalPath, string modPath, string targetPath, Action<string>? log = null)
        {
            try
            {
                if (!File.Exists(originalPath) || !File.Exists(modPath)) return false;

                var originalData = await File.ReadAllBytesAsync(originalPath);
                var modData = await File.ReadAllBytesAsync(modPath);

                var originalCrc = ComputeCrc32(originalData);
                var modCrc = ComputeCrc32(modData);

                log?.Invoke($"Original CRC: 0x{originalCrc:X8}, Mod CRC: 0x{modCrc:X8}");

                if (modCrc == originalCrc)
                {
                    await File.WriteAllBytesAsync(targetPath, modData);
                    log?.Invoke("CRC match, copied mod without patch.");
                    return true;
                }

                var modifiedCrcWithPadding = ComputeCrc32(new ReadOnlySpan<byte>(Concat(modData, new byte[4])));

                var originalBytes = UInt32ToBytesBE(originalCrc);
                var modifiedBytes = UInt32ToBytesBE(modifiedCrcWithPadding);

                var xorResult = XorBytes(originalBytes, modifiedBytes);
                var reversedBytes = ReverseBitsInBytes(xorResult);
                var k = new BigInteger(BytesToUInt32BE(reversedBytes));

                var crc32Poly = new BigInteger(0x104C11DB7);
                var correctionValue = GfInverse(k, crc32Poly);

                var correctionBytesRaw = UInt32ToBytesBE(correctionValue);
                byte ReverseBits(byte b)
                {
                    byte r = 0;
                    for (int i = 0; i < 8; i++)
                    {
                        if (((b >> i) & 1) == 1)
                            r |= (byte)(1 << (7 - i));
                    }
                    return r;
                }
                var correctionBytes = new byte[4];
                for (int i = 0; i < 4; i++)
                {
                    correctionBytes[i] = ReverseBits(correctionBytesRaw[i]);
                }

                var finalData = Concat(modData, correctionBytes);
                var finalCrc = ComputeCrc32(finalData);

                if (finalCrc == originalCrc)
                {
                    await File.WriteAllBytesAsync(targetPath, finalData);
                    log?.Invoke($"Patched CRC matched original (0x{finalCrc:X8}).");
                    return true;
                }

                log?.Invoke($"CRC patch failed. Final CRC: 0x{finalCrc:X8}, Expected: 0x{originalCrc:X8}");
                return false;
            }
            catch (Exception ex)
            {
                log?.Invoke($"CRC manipulation failed: {ex.Message}");
                return false;
            }
        }

        private static byte[] Concat(byte[] first, byte[] second)
        {
            var result = new byte[first.Length + second.Length];
            Buffer.BlockCopy(first, 0, result, 0, first.Length);
            Buffer.BlockCopy(second, 0, result, first.Length, second.Length);
            return result;
        }
    }
}
